### ISSUE TITLE
PHP 8 Support

### ISSUE THREAD

**Issue by Tastaturberuf** [contao-community-alliance/composer-plugin#83]
Reicht es die composer.json anzupassen? Dann mach ich gerne PR. 😂

**Comment by fritzmg** [comment]
Just test it with `--ignore-platform-req=php` ;)

**Comment by discordier** [comment]
I'd love to have this plugin phased out... when moving to PHP 8, one should finally have changed the Extension to a proper bundle instead of still relying on Contao 3.5 behaviour.

I mean, for most extensions it is a matter of:
- Shuffle files around
- Add the plugin
- remove autoload.ini

This would clear a bunch of deprecations in Contao then...

**Comment by Tastaturberuf** [comment]
@discordier Mit Contao 5, sind wir uns alle einig, brauchen wir das nicht mehr supporten. Vielleicht auch vorher, das sollte aber an die Contao-Version und nicht an die Platform gekoppelt sein. Es macht aber keinen Sinn, weil es mit PHP 7 und der gleichen Contao-Version eben geht.

**Comment by discordier** [comment]
This then still depends on https://github.com/contao/contao/pull/2291 which seems stuck for the moment - you see, it works with PHP and the same Contao version but NOT with PHP 8 and the same Contao version.

As I see some considerable amount of workload in Contao core still to be done to make that one work. I simply feel it won't be worth the hassle to bump this one also to PHP 8.
While it should definitely be possible to make it work with PHP 8, I doubt that the extensions will work "as is" in PHP 8 and as long as they have to be touched anyway, it seems more easy to make proper bundles from then than add yet more monkey patches to keep the legacy layout alive.
It's simply a matter of efficiency in the resulting code base in favor over being efficient (meaning: being too lazy to update the own code now and do it at Contao 5).
There are simply a huge load of features in Contao 4.x that get crippled by still using legacy extensions which is why I emphase they should simply not be used anymore. A ton of deprecation notices being generated every page impression that eat CPU cycles making your page slower. Service loading and DI not possible via legacy etc.

I simply wanted to state that I consider legacy extensions a bad idea, given that Contao 4.0 was released on 09.06.2015 and therefore all of you had 5 years to phase the legacy extensions out or reformat them to the real thing. The legacy compatibility was never meant to be an alternative but merely as a helper to ease migration to the new layout. Having it for 5 years now, and being even extended over all PHP 7 versions, do we really want to extend it to PHP 8 now?

But on the other hand, if you are willing to add it and provide support, feel free to do so, I won't stop you. :smiley_cat:

**Comment by dmolineus** [comment]
The Contao 4 support of the plugin was intended for the migration to Contao 4 structure. If you want to upgrade a Contao project to **new** platform requirements, you should also upgrade the project accordingly. I agree with @discordier that we from the Contao Community Alliance will not invest time in a dying technology.

If someone wants to achieve PHP 8 compatibility of his project by updating the plugin, we have nothing about a fork or open the plugin for new maintainers.

**Comment by aschempp** [comment]
I've discussed this with the Contao core team. There are plenty of Contao 3.x extensions that are still maintained for Contao 4, including most of the top @terminal42 modules (e.g. notification center). We (@terminal42) usually do not just restructure for Contao 4, because that would be an unnecessary BC break for existing users/devs (there are >10 extensions to Notification Center: https://extensions.contao.org/?q=notification%20center&pages=1).

If you're not heavily against it, I'll bump the plugin v3 to PHP 8, and I'm happy to (continue to) maintain it 😊

**Comment by discordier** [comment]
As stated above, we are not heavily against it but argue that there are should not be bc breaks when simply restructuring (introducing a bundle class and dropping the non functional legacy autoloader).
Keeping them as Contao 3 modules merely simply causes a ton of deprecation messages.

It would be interesting to know the views of the Contao core team on the matters of what is recommended.
Currently it sends the signal that legacy modules are not only "supported" but it is also "encouraged to not modernize them for bc reasons".
This however also implies to only modernize when bumping the major version which (we know from history of extensions) will not work out well as no one will do it.

To be honest and without any shaming intended, the fact that terminal42 extensions are NOT modernized was the reason (e.g. notification_center) were ruled out for some projects by the customer as they cause unremovable deprecation notices and appeared "abandoned" - in the means of not fully working with the current Contao version.

Long story short, feel free to bump to V3 and keep it alive - for me Contao 3 and it's extension loading mechanism is dead.

**Comment by aschempp** [comment]
Well if I move legacy classes to a bundle, I would need to (or am highly encouraged to) adjust the namespace => BC break? Also, what deprecation messages do you get that you wouldn't get after restructuring?

**Comment by aschempp** [comment]
Updated in bddeff9af32b7f271afb7814d8c94a80fe7b32c8
